torchtune/utils/_distributed.py

```diff
@@ -297,3 +299,31 @@ def lora_wrap_fsdp(module: nn.Module, recurse: bool, **kwargs):
         return isinstance(module, tuple(modules_to_wrap))
 
     return lora_wrap_fsdp
+
+
+def load_from_full_state_dict(
+    model: torch.distributed._composable.fsdp.FSDP,
+    full_sd: Dict[str, Any],
+    device: torch.device,
+    is_rank_zero: bool,
+):
+    meta_sharded_sd = model.state_dict()
+    sharded_sd = {}
+    for param_name, full_tensor in full_sd.items():
+        sharded_meta_param = meta_sharded_sd.get(param_name)
+        mesh = sharded_meta_param.device_mesh
+        if is_rank_zero:
+            full_tensor = full_tensor.detach().to(device)
+            torch.distributed.broadcast(full_tensor, src=0, group=mesh.get_group(0))
```

Comment:
I realized that `mesh.get_group(0)` is not right if we want this to generalize to 2D (just for future-proofing).

I might be missing something, but it looks like `DeviceMesh` does not have the API for what we want 🤔 
In the 2D case, we want to get the global PG over both TP and DP ranks, but it looks like `get_group()` would return a list of the DP PG and TP PG, meaning that we would need to make one `broadcast` call per mesh dim.

cc: @wanchaol @wz337 Is there any way to get the single global PG from a (parent) `DeviceMesh`?

Reply:
hmmm why we want the global pg for FSDP broadcasting? I think FSDP broadcast states would only require dp_pg to make sure every rank in the DP group starting off the same

Reply:
For this particular case, we are broadcasting the full state dict (source of truth) from rank 0 to all other ranks. We then shard the replicated tensor via `distribute_tensor`.

Reply:
The broadcasting is not specific to FSDP but rather to full state dict -> sharded state dict conversion. If we wanted to convert a full state dict on rank 0 to a sharded state dict on all ranks for a pure TP model, we can use this same utility.

Reply:
Aha I see, so only rank0 got the full_state_dict initially instead of every rank, that's why we need to do a whole world broadcasting?

Reply:
Ok not yet sure why we don't load full state dict to every rank but if rank0 only, we can just use the `get_default_group` API from c10d to get a world_pg?

Even if DeviceMesh offers such a thing, it would be just calling into this function I feel. Given that we are not sure yet if we want to expose a default_pg API in DeviceMesh (maybe we can have a `get_world_group` API, but I feel it's a bit over kill if c10d already provide such a thing)

Reply:
> Ok not yet sure why we don't load full state dict to every rank

1. We do not want to 8x the CPU RAM usage. If we `torch.load(mmap=True)` can allow processes to share the CPU memory, then this is not a problem.
2. We may not want to 8x the H2D copy volume, which may be slower than broadcast (depends on network bandwidth).
    - Each rank needs to copy each replicated tensor from CPU to GPU. Today, `distribute_tensor()` moves the tensor to device before sharding, but in theory, the sharding could be done on CPU too, meaning that this is not a fundamental problem.

For (1), we can explore if `torch.load(mmap=True)` can allow processes to share the CPU memory, but for (2), we need to benchmark to understand the performance implication for different setups.

We can discuss the option to load full state dict on all ranks more offline. I have some more thoughts.

---

Using the global PG seems okay for now. @weifengpy we can just not pass `group` then, and `dist.broadcast` will use the global PG by default.

Reply:
I see, yeah this make sense to me!

Reply:
removed `group` argument from `dist.broadcast` to use global PG